chall10/gbudau.py

```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/env python3
+
+import fileinput
+import sys
+import curses.ascii
+
+POSITION = "ABCDEFG"
+TEAMS = [1, 2]
+NOT_CONNECTED = 3
+CONNECTED = 4
+ROWS = 6
+COLUMNS = 7
+
+def init_board():
+    board = [[0] * COLUMNS for i in range(COLUMNS)]
+    return(board)
+
+def valid_input(action):
+    if not action or len(action) < 3 or action[0] not in POSITION or \
+       action[1] not in "_" or not curses.ascii.isupper(action[2]):
+        return False
+    return True
+
+def check_horizontal(board, team, row, col):
+    connect = 0
+    for i in board[row][col:]:
+        if i == TEAMS[team]:
+            connect += 1
+        else:
+            break
+    for i in board[row][:col][::-1]:
+        if i == TEAMS[team]:
+            connect += 1
+        else:
+            break
+    return team if connect >= CONNECTED else NOT_CONNECTED
+
+def check_vertical(board, team, row, col):
+    connect = 0
+    for i in range(1, row)[::-1]:
+        if board[i][col] == TEAMS[team]:
+            connect += 1
+        else:
+            break
+    for i in range(row, ROWS + 1):
+        if board[i][col] == TEAMS[team]:
+            connect += 1
+        else:
+            break
+    return team if connect >= CONNECTED else NOT_CONNECTED
+
+def check_diagonal_nesw(board, team, row, col):
+    connect = 0
+    j = col
+    for i in range(1, row)[::-1]:
+        j += 1
+        if j < COLUMNS and board[i][j] == TEAMS[team]:
+            connect += 1
+        else:
+            break
+    j = col
+    for i in range(row, ROWS + 1):
+        if j >= 0 and board[i][j] == TEAMS[team]:
+            connect += 1
+            j -= 1
+        else:
+            break
+    return team if connect >= CONNECTED else NOT_CONNECTED
+
+def check_diagonal_nwse(board, team, row, col):
+    connect = 0
+    j = col
+    for i in range(1, row)[::-1]:
+        j -= 1
+        if j >= 0 and board[i][j] == TEAMS[team]:
+            connect += 1
+        else:
+            break
+    j = col
+    for i in range(row, ROWS + 1):
+        if j < COLUMNS and board[i][j] == TEAMS[team]:
+            connect += 1
+            j += 1
+        else:
+            break
+    return team if connect >= CONNECTED else NOT_CONNECTED
+
+def check_diagonals(board, team, row, col):
+    connected = check_diagonal_nesw(board, team, row, col)
+    if connected != NOT_CONNECTED:
+        return team
+    connected = check_diagonal_nwse(board, team, row, col)
+    if connected != NOT_CONNECTED:
+        return team
+    return NOT_CONNECTED
+
+def check_connections(board, team, row, col):
```

Comment:
I think there's a lot of overhead with this way of approaching the problem, you try to `return` the answer to your `main()` function, when you could have simply `sys.exit(f"{winner}.")` in the `check_horizontal()`, check_vertical()` or check_diagonals()` functions.

Also I don't really like the `NOT_CONNECTED` variable, it's a bit confusing.

Reply:
I prefered to return an error code from those functions because it was easier for me to test them.
Could have used -1 instead of NOT_CONNECTED, it is a bit confusing :+1: